File: gym/kadanes_algorithm.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>

using namespace std;

int kadane(const vector<int> &a) {
  int max_so_far = numeric_limits<int>::min();
  int max_ending_here = 0;

  for (int e : a) {
    max_ending_here += e;
    max_so_far = max(max_so_far, max_ending_here);

    if (max_ending_here < 0) {
      max_ending_here = 0;
    }
  }

  return max_so_far;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;
    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    cout << kadane(a) << "\n";
  }

  return 0;
}

File: gym/KadanesAlgorithm.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class KadanesAlgorithm {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        System.out.printf("%d\n", kadane(a));
      }
    }
  }

  private static int kadane(int[] a) {
    int maxSoFar = Integer.MIN_VALUE;
    int maxEndingHere = 0;

    for (int e : a) {
      maxEndingHere += e;
      maxSoFar = Math.max(maxSoFar, maxEndingHere);

      if (maxEndingHere < 0) {
        maxEndingHere = 0;
      }
    }

    return maxSoFar;
  }
}